File: backend/services/priorityEngine.js
// Configurable Weights & Max Scores
const CONFIG = {
  weights: {
    safetyRisk: 30,
    publicImpact: 20,
    essentialService: 20,
    urgency: 10,
    duplicates: 10,
    location: 5,
    timePending: 5
  },
  safetyRiskMap: {
    "High": 30,
    "Critical": 30,
    "Medium": 15,
    "Moderate": 15,
    "Low": 5,
    "Minor": 5
  },
  publicImpactMap: {
    "High": 20,
    "Medium": 10,
    "Low": 5
  },
  urgencyMap: {
    "High": 10,
    "Medium": 5,
    "Low": 2
  },
  locationKeywords: [
    "hospital", "clinic", "school", "college", "station", "highway",
    "market", "airport", "bus stand", "bus stop", "metro", "railway"
  ]
};

/**
 * Computes priority score, level, breakdown, and explanation.
 * 
 * @param {Object} params
 * @param {string} params.description
 * @param {string} params.address
 * @param {number} params.duplicateCount
 * @param {string} params.createdAt - ISO Date string
 * @param {string} params.status
 * @param {Object} params.aiMetadata - { safetyRisk, publicImpact, essentialService, urgency }
 * 
 * @returns {Object} { priorityScore, priorityLevel, priorityBreakdown, reason }
 */
function calculatePriority({
  description = "",
  address = "",
  duplicateCount = 0,
  createdAt = null,
  status = "Submitted",
  aiMetadata = {}
}) {
  // 1. Safety Risk
  const safetyVal = aiMetadata.safetyRisk || "Medium";
  const safetyScore = CONFIG.safetyRiskMap[safetyVal] !== undefined ? CONFIG.safetyRiskMap[safetyVal] : 15;

  // 2. Public Impact
  const impactVal = aiMetadata.publicImpact || "Medium";
  const impactScore = CONFIG.publicImpactMap[impactVal] !== undefined ? CONFIG.publicImpactMap[impactVal] : 10;

  // 3. Essential Service
  const essentialVal = !!aiMetadata.essentialService;
  const essentialScore = essentialVal ? CONFIG.weights.essentialService : 0;

  // 4. Urgency
  const urgencyVal = aiMetadata.urgency || "Medium";
  const urgencyScore = CONFIG.urgencyMap[urgencyVal] !== undefined ? CONFIG.urgencyMap[urgencyVal] : 5;

  // 5. Duplicate Complaints
  const duplicatesScore = Math.min(CONFIG.weights.duplicates, duplicateCount * 5);

  // 6. Location Importance
  let locationScore = 0;
  const combinedText = `${description} ${address || ""}`.toLowerCase();
  if (CONFIG.locationKeywords.some(kw => combinedText.includes(kw))) {
    locationScore = CONFIG.weights.location;
  }

  // 7. Time Pending
  let timePendingScore = 0;
  if (status !== "Resolved" && createdAt) {
    const elapsedMs = Date.now() - new Date(createdAt).getTime();
    const elapsedHours = elapsedMs / (1000 * 60 * 60);
    // 1 point per 24 hours pending, capped at 5
    timePendingScore = Math.min(CONFIG.weights.timePending, Math.floor(elapsedHours / 24));
  }

  // Total Score
  let totalScore = safetyScore + impactScore + essentialScore + urgencyScore + duplicatesScore + locationScore + timePendingScore;
  totalScore = Math.max(0, Math.min(100, totalScore));

  // Determine Level
  let level = "Medium";
  if (totalScore >= 81) level = "Critical";
  else if (totalScore >= 61) level = "High";
  else if (totalScore >= 31) level = "Medium";
  else level = "Low";

  // Formulate Reason
  const reasons = [];
  if (safetyScore >= 15) reasons.push(`a safety risk flagged as ${safetyVal}`);
  if (essentialVal) reasons.push("disruption to essential services");
  if (duplicatesScore > 0) reasons.push(`linked to multiple similar reports (${duplicateCount} duplicate(s) detected)`);
  if (locationScore > 0) reasons.push("near critical public location/infrastructure");

  const reason = reasons.length > 0 
    ? `Complaint prioritized due to: ${reasons.join(", and ")}.`
    : `Complaint classified with general parameters. Priority level: ${level}.`;

  return {
    priorityScore: totalScore,
    priorityLevel: level,
    priorityBreakdown: {
      safetyRisk: safetyScore,
      publicImpact: impactScore,
      essentialService: essentialScore,
      urgency: urgencyScore,
      duplicates: duplicatesScore,
      location: locationScore,
      timePending: timePendingScore
    },
    reason
  };
}

module.exports = {
  calculatePriority,
  CONFIG
};

File: backend/app/services/priority_engine.py
import re
from datetime import datetime
from sqlalchemy.orm import Session
from app.models.complaint import Complaint

# Configurable Weights/Max Scores
MAX_SAFETY_RISK = 30
MAX_PUBLIC_IMPACT = 20
MAX_ESSENTIAL_SERVICE = 20
MAX_URGENCY = 10
MAX_DUPLICATES = 10
MAX_LOCATION = 5
MAX_TIME_PENDING = 5
MAX_WOMEN_SAFETY = 15  # boost for women-safety related grievances

# Women-safety detection: matched against category + description + address.
# These complaints get a dedicated priority boost so they surface first.
WOMEN_SAFETY_KEYWORDS = [
    "harass", "stalk", "eve teasing", "eve-teasing", "molest", "lewd",
    "women safety", "unsafe for women", "catcall", "followed me",
    "inappropriate touch", "chain snatch", "snatching", "robbery",
    "kidnap", "assault", "unsafe at night", "drunk men", "misbehav",
]

# Mapping Factors
SAFETY_RISK_MAP = {
    "High": 30,
    "Critical": 30,
    "Medium": 15,
    "Moderate": 15,
    "Low": 5,
    "Minor": 5
}

PUBLIC_IMPACT_MAP = {
    "High": 20,
    "Medium": 10,
    "Low": 5
}

URGENCY_MAP = {
    "High": 10,
    "Medium": 5,
    "Low": 2
}

LOCATION_KEYWORDS = [
    "hospital", "clinic", "school", "college", "station", "highway",
    "market", "airport", "bus stand", "bus stop", "metro", "railway"
]

def calculate_priority(
    db: Session,
    complaint_id: int,
    description: str,
    address: str,
    category: str,
    created_at: datetime,
    status: str,
    ai_metadata: dict,
    is_escalated: bool = False
    ai_metadata: dict,
    is_escalated: bool = False
) -> dict:
    """
    Computes a transparent rule-based priority score between 0 and 100.
    Returns:
        {
            "priorityScore": int,
            "priorityLevel": str,
            "priorityBreakdown": dict,
            "reason": str
        }
    """
    # 1. Safety Risk
    safety_risk_val = ai_metadata.get("safetyRisk", "Medium")
    safety_score = SAFETY_RISK_MAP.get(safety_risk_val, 15)
    
    # 2. Public Impact
    public_impact_val = ai_metadata.get("publicImpact", "Medium")
    impact_score = PUBLIC_IMPACT_MAP.get(public_impact_val, 10)
    
    # 3. Essential Service
    essential_val = ai_metadata.get("essentialService", False)
    essential_score = MAX_ESSENTIAL_SERVICE if essential_val else 0
    
    # 4. Urgency
    urgency_val = ai_metadata.get("urgency", "Medium")
    urgency_score = URGENCY_MAP.get(urgency_val, 5)
    
    # 5. Duplicate Complaints (same category and address or within same area)
    duplicate_count = 0
    if category and address and db:
        try:
            duplicate_count = (
                db.query(Complaint)
                .filter(
                    Complaint.category == category,
                    Complaint.address == address,
                    Complaint.id != complaint_id
                )
                .count()
            )
        except Exception as e:
            print(f"[Priority Engine] Error counting duplicates: {e}")
            
    # Score duplicate complaints: 5 points for 1 duplicate, 10 points for 2+ duplicates
    duplicates_score = min(MAX_DUPLICATES, duplicate_count * 5)
    
    # 6. Location Importance
    location_score = 0
    combined_text = (description + " " + (address or "")).lower()
    if any(kw in combined_text for kw in LOCATION_KEYWORDS):
        location_score = MAX_LOCATION
        
    # 7. Time Pending
    time_pending_score = 0
    if status not in ["Resolved", "Closed"] and created_at:
    if status not in ["Resolved", "Closed"] and created_at:
        hours_pending = (datetime.now() - created_at).total_seconds() / 3600
        # 1 point per 24 hours pending, capped at 5
        time_pending_score = min(MAX_TIME_PENDING, int(hours_pending // 24))

    # 8. SLA Escalation Boost (+20 points)
    escalation_score = 20 if is_escalated else 0

    # 9. Women Safety Boost (+15 points)
    # Detected from the AI category (Women Safety Cell) or safety keywords in
    # the complaint text — these grievances must never sit at the bottom.
    category_l = (category or "").lower()
    women_safety_score = 0
    if (
        "eve teasing" in category_l
        or "unsafe spot" in category_l
        or "women" in category_l
        or any(kw in combined_text for kw in WOMEN_SAFETY_KEYWORDS)
    ):
        women_safety_score = MAX_WOMEN_SAFETY

    # Total Score Calculation
    total_score = (
        safety_score +
        impact_score +
        essential_score +
        urgency_score +
        duplicates_score +
        location_score +
        time_pending_score +
        escalation_score +
        women_safety_score
    )
    total_score = max(0, min(100, total_score))
    
    # Determine Priority Level
    if total_score >= 81:
        level = "Critical"
    elif total_score >= 61:
        level = "High"
    elif total_score >= 31:
        level = "Medium"
    else:
        level = "Low"
        
    # Build Reason Explanation
    reasons = []
    if safety_score >= 15:
        reasons.append(f"a safety risk flagged as {safety_risk_val}")
    if essential_val:
        reasons.append("disruption to essential services")
    if duplicates_score > 0:
        reasons.append(f"linked to multiple similar reports ({duplicate_count} duplicate(s) detected)")
    if location_score > 0:
        reasons.append("near critical public location/infrastructure")
    if is_escalated:
        reasons.append("SLA breach escalation boost")
    if women_safety_score > 0:
        reasons.append("flagged as a public-safety concern (dedicated priority boost applied)")

    if reasons:
        reason_explanation = f"Complaint prioritized due to: {', and '.join(reasons)}."
    else:
        reason_explanation = f"Complaint classified with general parameters. Priority level: {level}."
        
    return {
        "priorityScore": total_score,
        "priorityLevel": level,
        "priorityBreakdown": {
            "safetyRisk": safety_score,
            "publicImpact": impact_score,
            "essentialService": essential_score,
            "urgency": urgency_score,
            "duplicates": duplicates_score,
            "location": location_score,
            "timePending": time_pending_score,
            "escalationBoost": escalation_score,
            "womenSafety": women_safety_score
        },
        "reason": reason_explanation
    }
